### PR TITLE
Use a custom ConcurrentQueue in debug mode for synthetic members

### PR DESCRIPTION
**Pull request by gafter**
to help detect possible concurrent addition of synthetic members to a type. Since we emit
synthesized members for each type in the order in which they appear in the queue, this might
detect some cases of nondeterminism.

I will add an issue to remove this code around October/November. Filed #4337 to track that.

@jaredpar @AlekseyTs @agocke @VSadov

### CODE REVIEW

File: src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs [hunk 1]
```diff
@@ -308,6 +308,52 @@ private ImmutableArray<Cci.AssemblyReferenceAlias> CalculateAssemblyReferenceAli
...
+        /// <typeparam name="T"></typeparam>
+        public class ConcurrentQueue<T> : System.Collections.Concurrent.ConcurrentQueue<T>
+        {
+            // A count of the number of concurrent queue operations in progress. Should always be zero or one,
+            // as synthetic members should be added by the compiler to a given type in a well-defined sequential
+            // order.
+            int queueing;
+
+            // A short delay to increase the chance that concurrent Enqueue operation will be diagnosed.
+            static TimeSpan shortDelay = new TimeSpan(ticks: ShouldDelay() ? 2 : 0);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs]: Readonly?

File: src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs [hunk 2]
```diff
@@ -308,6 +308,52 @@ private ImmutableArray<Cci.AssemblyReferenceAlias> CalculateAssemblyReferenceAli
...
+
+            // The delay is set to change to zero after October 2015, when this whole class should be removed.
+            static bool ShouldDelay()
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs]: I'd rather we file a bug to track this than add a time bomb to the code.

File: src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs [hunk 3]
```diff
@@ -308,6 +308,52 @@ private ImmutableArray<Cci.AssemblyReferenceAlias> CalculateAssemblyReferenceAli
...
         #region Synthesized Members
 
+#if DEBUG
+        /// <summary>
+        /// The queue of synthesized members of a type should be deterministic, as the members are
+        /// emitted in the order in which they are added to the queue. Therefore for debug purposes
+        /// we have a custom version of ConcurrentQueue that detects attempted concurrent adds.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        public class ConcurrentQueue<T> : System.Collections.Concurrent.ConcurrentQueue<T>
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs]: Can this be private?

Comment (gafter) [inline src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs]: Sure. The enclosing type is already internal.

File: src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs]: OK, sure.

File: src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs]: Per @jaredpar 's comment I will remove this.

File: src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs [hunk 4]
```diff
@@ -308,6 +308,52 @@ private ImmutableArray<Cci.AssemblyReferenceAlias> CalculateAssemblyReferenceAli
...
+            public new void Enqueue(T item)
+            {
+                if (Interlocked.Increment(ref queueing) != 1)
+                {
+                    throw new System.InvalidOperationException("Concurrent use of " + nameof(SynthesizedDefinitions));
+                }
+                base.Enqueue(item);
+                // To increase the chance of catching concurrency issues, we add a delay to each queued item
+                // so that another thread has a chance to add at the same time.
+                System.Threading.Tasks.Task.Delay(shortDelay).Wait();
```
Comment (VSadov) [inline src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs]: Could this be just   Thread.Sleep(shortDelay)  ?

Comment (gafter) [inline src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs]: No, the portable platform does not have the type Thread.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by VSadov** [comment]
LGTM

**Pull request closed without merging**